File: gateway/src/services/normalize.ts
/**
 * URL 规范化
 * 保证同一内容的不同链接形式能被准确识别
 */

/** 需要移除的追踪参数 */
const TRACKING_PARAMS = new Set([
  'utm_source', 'utm_medium', 'utm_campaign', 'utm_term', 'utm_content',
  'ref', 'source', 'spm', 'from', 'share_source', 'share_medium',
  'fbclid', 'gclid', 'msclkid', 'mc_cid', 'mc_eid',
]);

/** 域名规范化映射 */
const DOMAIN_ALIASES: Record<string, string> = {
  'x.com': 'twitter.com',
  'www.twitter.com': 'twitter.com',
  'mobile.twitter.com': 'twitter.com',
  'www.github.com': 'github.com',
  'www.zhihu.com': 'zhihu.com',
  'zhuanlan.zhihu.com': 'zhihu.com/p',
};

/**
 * 规范化 URL
 */
export function normalize(rawUrl: string): string {
  let url: URL;
  
  try {
    url = new URL(rawUrl);
  } catch {
    // 如果解析失败，返回原始 URL
    return rawUrl;
  }
  
  // 1. scheme 和 host 转小写
  let hostname = url.hostname.toLowerCase();
  
  // 2. 域名别名规范化
  if (DOMAIN_ALIASES[hostname]) {
    hostname = DOMAIN_ALIASES[hostname];
  }
  
  // 3. 移除 www. 前缀（除非已经在映射中处理）
  if (hostname.startsWith('www.') && !DOMAIN_ALIASES[hostname]) {
    hostname = hostname.slice(4);
  }
  
  // 4. 规范化路径
  let pathname = url.pathname;
  
  // 移除末尾斜杠（除非是根路径）
  if (pathname.length > 1 && pathname.endsWith('/')) {
    pathname = pathname.slice(0, -1);
  }
  
  // GitHub 特殊处理：移除 /tree/main, /tree/master 等默认分支路径
  if (hostname === 'github.com') {
    pathname = pathname.replace(/\/tree\/(main|master)$/, '');
    pathname = pathname.replace(/\/blob\/(main|master)\//, '/blob/main/');
  }
  
  // 5. 过滤查询参数
  const params = new URLSearchParams();
  for (const [key, value] of url.searchParams.entries()) {
    if (!TRACKING_PARAMS.has(key.toLowerCase())) {
      params.append(key, value);
    }
  }
  
  // 6. 排序查询参数
  params.sort();
  
  // 7. 重建 URL（不包含 fragment）
  const queryString = params.toString();
  const normalizedUrl = `https://${hostname}${pathname}${queryString ? '?' + queryString : ''}`;
  
  return normalizedUrl;
}

/**
 * 从 URL 推断 source
 */
export function inferSource(url: string): string {
  try {
    const parsed = new URL(url);
    const hostname = parsed.hostname.toLowerCase();
    
    if (hostname.includes('github.com')) return 'github';
    if (hostname.includes('twitter.com') || hostname.includes('x.com')) return 'twitter';
    if (hostname.includes('zhihu.com')) return 'zhihu';
    if (hostname.includes('xiaohongshu.com') || hostname.includes('xhslink.com')) return 'xiaohongshu';
    
    return 'others';
  } catch {
    return 'others';
  }
}

File: collector/normalize.py
"""
DEPRECATED (2026-04): This module is being replaced by Gateway API.

The Gateway service now handles URL normalization via:
- gateway/src/utils/normalize.ts

This module is kept for backward compatibility but will be removed in a future version.
New collectors should rely on Gateway's built-in normalization.

Migration guide:
- Old: from collector.normalize import normalize_url
- New: Gateway normalizes URLs automatically in /api/v1/collect endpoint
"""

from urllib.parse import urlparse, urlunparse, parse_qs, urlencode
import re
from typing import Dict, Any

PLATFORM_RULES = {
    'github': {
        # https://github.com/owner/repo -> https://github.com/owner/repo
        # https://github.com/owner/repo/tree/main -> https://github.com/owner/repo
        'pattern': r'^(https://github\.com/[^/]+/[^/]+).*$',
        'replace': r'\1'
    },
    'twitter': {
        'alias': ['x.com'],
        'canonical': 'twitter.com'
    },
    'xiaohongshu': {
        'strip_www': True
    }
}

def normalize_url(url: str, platform=None) -> str:
    """
    Standardize a URL to prevent duplicate collections of the same content.
    
    Basic Rules:
    1. Lowercase scheme and netloc
    2. Remove trailing slash
    3. Remove tracking parameters
    4. Remove URL fragments
    5. Sort query parameters alphabetically
    """
    parsed = urlparse(url)
    
    # 1. Scheme + netloc to lowercase
    scheme = parsed.scheme.lower()
    netloc = parsed.netloc.lower()
    
    # 2. Remove trailing slash
    path = parsed.path.rstrip('/')
    if not path:
        path = ''
        
    # 3. Filter tracking parameters
    query_params = parse_qs(parsed.query, keep_blank_values=True)
    tracking_params = {
        'utm_source', 'utm_medium', 'utm_campaign', 'utm_term', 
        'utm_content', 'ref', 'source', 'from', 'spm', 'fbclid'
    }
    filtered_params = {
        k: v for k, v in query_params.items() 
        if k.lower() not in tracking_params
    }
    
    # 4. Remove fragment
    fragment = ''
    
    # 5. Sort query params
    sorted_query = urlencode(sorted(filtered_params.items()), doseq=True)
    
    # Rebuild basic normalized URL
    normalized = urlunparse((scheme, netloc, path, '', sorted_query, fragment))
    
    # Apply platform-specific rules
    if platform and platform in PLATFORM_RULES:
        rules = PLATFORM_RULES[platform]
        
        # Apply regex replacement
        if 'pattern' in rules and 'replace' in rules:
            normalized = re.sub(rules['pattern'], rules['replace'], normalized)
            
        # Handle domain aliases
        if 'alias' in rules and 'canonical' in rules:
            for alias in rules['alias']:
                if netloc.endswith(alias):
                    normalized = normalized.replace(alias, rules['canonical'], 1)
                    
        # Remove www prefix
        if 'strip_www' in rules and rules['strip_www']:
            if netloc.startswith('www.'):
                normalized = normalized.replace('www.', '', 1)
                
    return normalized
